### PR TITLE
[Merged by Bors] - refactor(AlgebraicGeometry/EllipticCurve/*): add `WeierstrassCurve.IsElliptic` and remove `EllipticCurve`

### PR DESCRIPTION
**Pull request by acmepjz**
The original `EllipticCurve` is `WeierstrassCurve` bundled with `Δ' : Rˣ` and a proof that `Δ' = Δ`. This definition makes `Δ'` and `j` computable, but at the expense that some APIs must be duplicated for `EllipticCurve` and `WeierstrassCurve`.

This PR adds the typeclass `WeierstrassCurve.IsElliptic` which asserts that `IsUnit W.Δ`, and which is a `Prop`. The old API for elliptic curves are replaced by a `WeierstrassCurve` `W` together with `[W.IsElliptic]`. The caveat is that `Δ'` and `j` becomes noncomputable.

The original design of this PR is `WeierstrassCurve.Elliptic` which is `Invertible W.Δ`, and which is a `Type`. By this design `Δ'` and `j` remain computable. But this design makes simpNF linter timeout. So this design is abandoned and replaced by current design.

Also rename `J.lean` -> `IsomOfJ.lean` and split `ModelsWithJ.lean`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

Discussions:

https://leanprover.zulipchat.com/#narrow/channel/116395-maths/topic/thoughts.20on.20elliptic.20curves/near/432354438

https://leanprover.zulipchat.com/#narrow/channel/116395-maths/topic/thoughts.20on.20elliptic.20curves/near/473423362

### CODE REVIEW

**Comment by acmepjz** [comment]
I've tested that `Elliptic : Type` -> `IsElliptic : Prop` and which eliminates the mysterious timeout error. At the expense that `Δ'` and `j` being noncomputable...

Note that the `IsInvertible` instance is still used, which contains data... Next step maybe we want to replaced them by `IsUnit` which contains no data, but for now `IsUnit` is not an instance... so I think this is related to #17458 ...

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (Multramate) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:525]: Shall we wait until #17458 is merged before continuing on this PR?

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (Multramate) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:585]: These and the lemmas below can be stated in fewer lines, since we no longer need `invertibleOfNonzero`.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (Multramate) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:660]: Ditto as above:
```suggestion
instance : Inhabited { W : WeierstrassCurve F // W.IsElliptic } :=
```

**Review by Multramate — COMMENTED** [review]
Thanks - this is wonderful and very welcome! I added a few comments and will continue with the reviews once I have more time in a few days.

One overarching question: do we want to stick with the variable `W : WeierstrassCurve` when we have `[W.IsElliptic]`? Should we instead use `E : WeierstrassCurve`?

File: Mathlib/AlgebraicGeometry/EllipticCurve/Affine.lean
Comment (acmepjz) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Affine.lean:17]: Please note that `Mathlib.AlgebraicGeometry.EllipticCurve.Group` is not working in this case; see https://leanprover-community.github.io/mathlib4_docs/Mathlib/AlgebraicGeometry/EllipticCurve/Affine.html .

> can you do the same for lines 14-15 in `Group.lean`?

Sure.

Comment (Multramate) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Affine.lean:17]: Yeah, I am very aware of this annoying bug in the docs! Do you have any idea why?

Comment (acmepjz) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Affine.lean:17]: According to https://github.com/leanprover/doc-gen4/blob/b6ae1cf11e83d972ffa363f9cdc8a2f89aaa24dc/DocGen4/Output/DocString.lean#L68-L100 the only correct way to reference file is write its filename, otherwise I don't expect it works (seems that it works in some cases, but I think it only works if the file referenced is already `import`ed by current file), since there is no symbol named `Mathlib.AlgebraicGeometry.EllipticCurve.Group`...

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (Multramate) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:351]: Yeah, that's fine too.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (Multramate) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:584]: Yeah that's what I'm thinking. The statements of the other lemmas can each be one line.

Comment (acmepjz) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean]: Done.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Affine.lean
Comment (acmepjz) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Affine.lean:17]: > can you do the same for lines 14-15 in `Group.lean`?

The line break in that file is too tight. 🤦‍♂️ Since it's working, I'll not change them.

Comment (Multramate) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Affine.lean:17]: I think this works?
```
under the geometric group law defined in `Mathlib.AlgebraicGeometry/EllipticCurve/Affine.lean` and
in `Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean`.
```

Comment (Multramate) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Affine.lean:17]: In any case, I guess don't worry too much about it because they will all be refactored eventually (hopefully soon).

**Comment by acmepjz** [comment]
> One overarching question: do we want to stick with the variable `W : WeierstrassCurve` when we have `[W.IsElliptic]`? Should we instead use `E : WeierstrassCurve`?

It doesn't matter for me. Mostly because that's just one `W : WeierstrassCurve` across whole file, and in some sections there are `variable [W.IsElliptic]`. If in the beginning it's required to be `IsElliptic`, then I think it's fine to use `E` instead (e.g. in `J.lean`).

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (Multramate) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:529]: Sorry, I just got back to this. Once that is merged, the suggestions I have below will be drastically different.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (Multramate) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:584]: Should we use `(h3 : (3 : F) ≠ 0)` instead? I don't quite remember why we did this instead.

Otherwise, we can also make the assumptions below `[CharP F 3]` etc, but we should be consistent.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (Multramate) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:594]:
```suggestion
  rw [ofJ, if_pos <| by linear_combination 576 * h3, if_pos h3]
```

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (Multramate) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:600]:
```suggestion
  · exact ofJ_1728_of_three_eq_zero h3
  · rw [ofJ, show (1728 : F) = 2 ^ 6 * 3 ^ 3 by norm_num1,
      if_neg <| mul_ne_zero (pow_ne_zero 6 h2.out) (pow_ne_zero 3 h3), if_pos rfl]
```

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (Multramate) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:603]:
```suggestion
  rw [ofJ, if_pos <| by linear_combination 864 * h2,
```

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (Multramate) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:612]: All the `haveI` can be replaced by `have`. Once we replace `Fact (IsUnit ...)` with `IsUnit ...`, and once we decide on whether to remove `NeZero`, most of these should be a lot cleaner to prove.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (Multramate) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:423]: With the prime linter, I think we should try to rename this? What do you think of calling one of them `disc`, and keep the other one `Δ`? The one that's used more often should be `Δ`, but I'm not sure which.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (Multramate) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:37]:
```suggestion
 * `WeierstrassCurve.IsElliptic`: typeclass asserting that a Weierstrass curve is an elliptic curve.
```

**Review by Multramate — COMMENTED** [review]
I wasn't aware of the existence of `J.lean`, so I'll mostly comment on `Weierstrass.lean`.

Do you think it could be better to move everything in `ModelsWithJ` there?

**Comment by acmepjz** [comment]
> I wasn't aware of the existence of `J.lean`

It's introduced in #17431.

> Do you think it could be better to move everything in `ModelsWithJ` there?

Note that that file imports `Mathlib.FieldTheory.IsSepClosed` as well as `Mathlib.AlgebraicGeometry.EllipticCurve.NormalForms`, which is not needed by `ModelsWithJ`.

Originally the `ModelsWithJ` is introduced to generalize `Inhabited (EllipticCurve Rat)` to any field. The original result is in `Weierstrass` so the `ModelsWithJ` is also in `Weierstrass`. But now since we don't have `EllipticCurve` anymore, so I think it's OK to move them to another file. Probably a new file `ModelsWithJ.lean` but not `J.lean`.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (acmepjz) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:423]: I don't know. Maybe we should leave this for another PR. Changing `Δ'` to something else requires altering some of existing code using `Δ'`.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (acmepjz) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:584]: > Should we use `(h3 : (3 : F) ≠ 0)` instead? I don't quite remember why we did this instead.

It's here https://github.com/leanprover-community/mathlib4/pull/5935#discussion_r1280637704

I think it's OK to revert to `(h3 : (3 : F) ≠ 0)` since otherwise it looks like that we are abusing instances.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (acmepjz) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:612]: I just get used to write `haveI` for instances instead of `have`. Anyways, I'll change them.

**Comment by acmepjz** [comment]
Done for most suggestions.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (Multramate) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:423]: Sure!

**Review by Multramate — APPROVED** [review]
Minor comments but I'm mostly happy with this. I'm not familiar with `J.lean` so I don't have anything to say. I guess my main question is: why the need to separate the single result of `J.lean` into a different file? I feel that `FieldTheory/IsSepClosed` is not too large of a dependency (perhaps comparable to `Algebra/CharP` and `RingTheory/Nilpotent` in `Weierstrass.lean`?). If the issue is the length of the file, then I can understand :P

@jcommelin back to you!

File: Mathlib/AlgebraicGeometry/EllipticCurve/VariableChange.lean
Comment (acmepjz) [reply Mathlib/AlgebraicGeometry/EllipticCurve/VariableChange.lean:50]: For humans this is easier to understand than the matrix multiplication?

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (acmepjz) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:413]: > I always thought that's mainly for name clashes.

Yes, I'm afraid that sometimes later we will have `PDE.IsElliptic`...

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (acmepjz) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:552]: Good idea since there are debates on `field_simp`.

File: Mathlib/AlgebraicGeometry/EllipticCurve/VariableChange.lean
Comment (Multramate) [reply Mathlib/AlgebraicGeometry/EllipticCurve/VariableChange.lean:51]: Right, then it should probably say something like:
```
In other words, this is the change of variables $(X, Y) \mapsto (u^2X + r, u^3Y + u^2sX + t)$.
```
or something else that fits in one line. It's implicit we're working with Weierstrass equations.

**Comment by Multramate** [comment]
> Anyways the file name `J.lean` maybe not the best, do you think `IsomOfJEq` or something else is better?

Yeah, that sounds better than `J.lean`. Maybe `IsomOfJ.lean` is clear enough.

**Review by jcommelin — COMMENTED** [review]
@Multramate Thanks for the reviews!

@acmepjz Thanks for the PR!

bors merge

**Pull request closed without merging**
